internal/smtpsrv/conn.go

```diff
@@ -1047,6 +1048,39 @@ func (c *Conn) AUTH(params string) (code int, msg string) {
 	response := ""
 	if len(sp) == 2 {
 		response = sp[1]
+	} else if sp[0] == "LOGIN" {
+		// With the LOGIN method, the user password and domain are
+		// passed in separate messages. Here we prompt for the LOGIN
+		// parameters and convert them into the PLAIN authentication
+		// format, i.e. the base64-encoded string:
+		//	<authorization id> NUL <authentication id> NUL <password>
+		if err := c.writeResponse(334, ""); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error writing AUTH 334: %v", err)
+		}
+		user := []byte{}
+		pass := []byte{}
+
+		if userb64, err := c.readLine(); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error reading AUTH LOGIN user response: %v", err)
+		} else if user, err = base64.StdEncoding.DecodeString(userb64); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error parsing AUTH LOGIN user 334: %v", err)
+		} else if err := c.writeResponse(334, ""); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error writing AUTH 334: %v", err)
+		}
+
+		if passb64, err := c.readLine(); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error reading AUTH LOGIN pass response: %v", err)
+		} else if pass, err = base64.StdEncoding.DecodeString(passb64); err != nil {
+			return 554, fmt.Sprintf("5.4.0 Error parsing AUTH LOGIN pass 334: %v", err)
+		}
+
+		plain := []byte{}
+		plain = append(plain, user...)
+		plain = append(plain, '\000')
+		plain = append(plain, user...)
+		plain = append(plain, '\000')
+		plain = append(plain, pass...)
+		response = base64.StdEncoding.EncodeToString(plain)
```

Comment:
Similar to the other comment I left, I think this would greatly benefit from splitting the function: it'd be easier and less error prone have the last part of this function be another function that takes `user, domain, passwd string` arguments.

Reply:
+1 I think this part can be ommitted if there is a refactor as mentioned in the other comment.

The helper functions can return user, domain and pass, and then AUTH can use that just like it does today.